source/agora/consensus/Quorum.d

```diff
@@ -0,0 +1,1229 @@
+}
+
+/*******************************************************************************
+
+    Verify that the provided quorum sets are considered sane by SCP.
+
+    The quorums are checked both pre and post-normalization,
+    with extra safety checks enabled.
+
+    Params:
+        quorums = the quorum map of (node => quorum) to verify
+
+    Throws:
+        AssertError if the quorum is not considered sane by SCP.
+
+*******************************************************************************/
+
+private void verifyQuorumsSanity (QuorumConfig[PublicKey] quorums)
+{
+    import scpd.scp.QuorumSetUtils;
+
+    foreach (key, quorum; quorums)
+    {
+        auto scp_quorum = toSCPQuorumSet(quorum);
+        const(char)* reason;
+
+        assert(isQuorumSetSane(scp_quorum, true, &reason),
+            format("Key %s with %s fails sanity check before normalization: %s",
+                    key.nice, quorum.toToml, reason.to!string));
+
+        normalizeQSet(scp_quorum);
+        assert(isQuorumSetSane(scp_quorum, true, &reason),
+            format("Key %s with %s fails sanity check after normalization: %s",
+                    key.nice, quorum.toToml, reason.to!string));
+    }
+}
+
+/*******************************************************************************
+
+    Verify that all the quorums intersect according to the quorum checker tool
+
+    Params:
+        quorums = the quorums to check
+
+    Throws:
+        AssertError if the quorum is not considered sane by SCP.
+
+*******************************************************************************/
+
+private void verifyQuorumsIntersect (QuorumConfig[PublicKey] quorums,
+    size_t line = __LINE__)
+{
+    import agora.utils.Log;
+    auto qm = QuorumTracker.QuorumMap.create();
+
+    foreach (key, quorum; quorums)
+    {
+        auto scp = toSCPQuorumSet(quorum);
+        auto scp_quorum = makeSharedSCPQuorumSet(scp);
+
+        auto hash = StellarHash(key);
+        auto scp_key = NodeID(hash);
+
+        qm[scp_key] = scp_quorum;
+    }
+
+    auto qic = QuorumIntersectionChecker.create(qm);
+    if (!qic.networkEnjoysQuorumIntersection())
+    {
+        CircularAppender().printConsole();
+        assert(0);
+    }
+
+    auto splits = qic.getPotentialSplit();
+
+    if (splits.first.length != 0 ||
+        splits.second.length != 0)
+    {
+        CircularAppender().printConsole();
+
+        writefln("Splits: first: %s second: %s",
+            splits.first[].map!(node_id => key_to_simple[PublicKey(node_id)]),
+            splits.second[].map!(node_id => key_to_simple[PublicKey(node_id)]));
+
+        writefln("Quorum: %s", quorums.nice(0));
+
+        try
+        {
+            assert(0, "Splits found");  // should not happen
+        }
+        catch (Throwable thr)
+        {
+            thr.line = line;
+            throw thr;
+        }
+    }
+}
+
+/// very simplistic way of reducing a 64-byte blob to an 8-byte seed
+private ulong hashToSeed (Hash hash)
```

Comment:
We could potentially replace this with libsodium's `crypto_shorthash`. But it needs a key, and the key should be hardcoded (we need the reduction to be reproducible given the same input).